app/models/ogi_pilot_session.rb

```diff
@@ -1,6 +1,128 @@
 class OgiPilotSession < ApplicationRecord
+
+  has_secure_password
   validates :end_point, presence: true
   validates :channel, presence: true
   validates :topic, presence: true
   validates :message_type, presence: true
+  validates :user_name, presence: true
+
+  attr_accessor :validation_messages
+
+  def provider_session_exists
+    return self.provider_session_id.present?
+  end
+
+  def consumer_session_exists
+    return self.consumer_session_id.present?
+  end
+
+  def get_consumer_provider_details
+    # This function returns a new provider and consumer object.
+    return { consumer_client: IsbmRestAdaptor::ConsumerPublication.new, provider_client: IsbmRestAdaptor::ProviderPublication.new }
+  end
+
+  def open_session
+    client_details = get_consumer_provider_details()
+    self.consumer_session_id = client_details[:consumer_client].open_session(self.channel, [self.topic])
+    self.provider_session_id = client_details[:provider_client].open_session(self.channel)
```

Comment:
I would probably have two separate methods, `consumer_client` and `provider_client` that use lazy initialisation to return the clients objects. That way the first time you call them it creates the client and stores it as an instance variable, and subsequent calls use the same object. That way the rest of the code that uses the client objects just call `consumer_client.open_session`, `consumer_client.read_message`, etc., which makes it much simpler and clearer to understand. It also means it does not have to instantiate both clients in cases where you only need to use one of them.